### ISSUE TITLE
Compatibility with python3

### ISSUE THREAD

**Issue by cscoccola** [brinckmann/montepython_public#98]
Dear all,
I would like to know if you are planning to upgrade montepython to run under python3, since python2.7 is not longer supported (https://pythonclock.org/). Thanks and regards.

**Comment by brinckmann** [comment]
Hi Claudia,

MontePython v3.3, which will be in Python 3, should be released very soon.

Best,
Thejs

**Comment by ivandebono** [comment]
> Dear all,
> I would like to know if you are planning to upgrade montepython to run under python3, since python2.7 is not longer supported (https://pythonclock.org/). Thanks and regards.

I made a Python-3 compatible version (tested for Python 3.6.7). It's in a private repository. You should have received the link. 

Depending on what functionalities you plan to use (Multinest, Polychord, etc.), you need to be careful about how you compile any external packages. CLASS itself needs to be compiled separately for Python 3.

**Comment by brinckmann** [comment]
Version 3.3 of MontePython is now public, featuring Python 3 compatibility (with Python 2.7 still supported).

Thanks Ivan, we already had an internal version that was waiting for testing to be finished, plus waiting on other features planned for v3.3.

Best,
Thejs

**Comment by ivandebono** [comment]
There's a possible bug in line 1840 of analyze.py:

```
  File "montepython-3.3/montepython/analyze.py", line 1840, in remove_bad_points
    for line in py_ifilter(iscomment,f):
NameError: name 'py_ifilter' is not defined
```
Should that be filter(iscomment,f), as in Montepython 3.2?

**Comment by brinckmann** [comment]
Thanks, it should be `py_filter(iscomment,f)`, i.e. no i before filter.

**Comment by brinckmann** [comment]
I pushed a new version fixing this unfortunate bug!

**Comment by ivandebono** [comment]
There's a small bug in line 2136 of analyze.py. 

This:

`exec(open(command_line.optional_plot_file, plot_file_vars).read())`

should be changed to:

`exec(open(command_line.optional_plot_file).read(), plot_file_vars)`

**Comment by antoniuspisa** [comment]
If this might help, for my particular configuration, I had to add the following changes.

likelihood_class.py:2500:
matrix = read_table(path).as_matrix().reshape((length, length))
was changed into:
`matrix = read_table(path).values.reshape((length, length))`

In io_mp.py, line 483 was changed into:
            `lines_found += self.read(1024).count('\n'.encode())`

In sampler.py, line 106 was changed into:
        `if chain_file.tail(1)[0].split('\t'.encode())[0] == '#':`

In sampler.py line 118 was changed into:
            `chain_file.tail(commented_line+1)[commented_line].split('\t'.encode())[i])`

Thanks a lot to everyone.

**Comment by brinckmann** [comment]
@schoeneberg What do you think, should we add the encode() changes proposed by @antoniuspisa ?

Best,
Thejs

**Comment by brinckmann** [comment]
Also all the other bugfixes have (finally!) been added to branch 3.3.

Best,
Thejs

**Comment by schoeneberg** [comment]
Dear @brinckmann ,

in general I do agree with @antoniuspisa about the changes. However, for the tail command we should strongly consider to instead to change the tail command itself to something like

In io_mp.py the line 489 was changed to
`line_list = list([line.decode('utf-8') for line in self.readlines()])`
(untested)

In any case, there would need to be a `.count(b'\n')` , or equivalently the `.count('\n'.encode())` in line 487 in io_mp.py, though the former is a bit shorter. I haven't had time yet to test the decode option extensively, especially related to the performance in python2 and python3. This is something that I am still planning on doing, but haven't found time to do yet.

Cheers,
Nils Sch.

--- 

PS: The above change proposed in io_mp.py would make these changes of @antoniuspisa unnecessary:

In sampler.py, line 106 was changed into:
`if chain_file.tail(1)[0].split('\t'.encode())[0] == '#':`

In sampler.py line 118 was changed into:
`chain_file.tail(commented_line+1)[commented_line].split('\t'.encode())[i])`

**Comment by dchooper** [comment]
Hi,

All python3 issues reported here have been addressed in version 3.4, which has now been released.

Best,
Deanna
